src/Interpreters/ConcurrentHashJoin.cpp

```diff
@@ -43,36 +46,45 @@ ConcurrentHashJoin::ConcurrentHashJoin(ContextPtr context_, std::shared_ptr<Tabl
         inner_hash_join->data = std::make_unique<HashJoin>(table_join_, right_sample_block, any_take_last_row_);
         hash_joins.emplace_back(std::move(inner_hash_join));
     }
-
 }
 
-bool ConcurrentHashJoin::addJoinedBlock(const Block & block, bool check_limits)
+bool ConcurrentHashJoin::addJoinedBlock(const Block & right_block, bool check_limits)
 {
-    Blocks dispatched_blocks = dispatchBlock(table_join->getOnlyClause().key_names_right, block);
+    Blocks dispatched_blocks = dispatchBlock(table_join->getOnlyClause().key_names_right, right_block);
 
-    std::list<size_t> pending_blocks;
-    for (size_t i = 0; i < dispatched_blocks.size(); ++i)
-        pending_blocks.emplace_back(i);
-    while (!pending_blocks.empty())
+    std::atomic<size_t> blocks_left = 0;
+    for (const auto & block : dispatched_blocks)
+    {
+        if (block)
+        {
+            ++blocks_left;
+        }
+    }
+
+    while (blocks_left > 0)
     {
-        for (auto iter = pending_blocks.begin(); iter != pending_blocks.end();)
+        /// insert blocks into corresponding HashJoin instances
+        for (size_t i = 0; i < dispatched_blocks.size(); ++i)
         {
-            auto & i = *iter;
             auto & hash_join = hash_joins[i];
             auto & dispatched_block = dispatched_blocks[i];
-            if (hash_join->mutex.try_lock())
             {
-                if (!hash_join->data->addJoinedBlock(dispatched_block, check_limits))
-                {
-                    hash_join->mutex.unlock();
-                    return false;
-                }
+                /// if current hash_join is already processed by another thread, skip it and try later
+                std::unique_lock<std::mutex> lock(hash_join->mutex, std::try_to_lock);
+                if (!lock.owns_lock())
+                    continue;
 
-                hash_join->mutex.unlock();
-                iter = pending_blocks.erase(iter);
+                if (!dispatched_block)
+                    continue;
```

Comment:
You can do this check before trying to acquire the lock

Reply:
Seems that it's possible that  another thread will do `dispatched_block = {}` between check and acquiring mutex.